### PR TITLE
add JS resolver support

### PR DESCRIPTION
**Pull request by iartemiev**
## Changes

The Amplify CDK construct does not currently support provisioning AppSync JS resolvers for fields defined in the customer's Amplify schema.

This PR leverages the AppSync L1 CDK construct escape hatch to add support. 

Related data-schema PR: https://github.com/aws-amplify/amplify-api-next/pull/113

**Corresponding docs PR, if applicable:** https://github.com/aws-amplify/docs/pull/6836/files#diff-dfe0a992e880882948237e578edffdfb4a4548f19487a58118ff900d2c7c233aR178-R210

### Customer DX

```ts
const schema = a.schema({
  Post: a.model({...}),

  likePost: a.mutation()
    .returns( a.string() )
    .authorization([ a.allow.private() ])
    .handler([
      // a.handler.custom is how customers define an AppSync JS Resolver
      a.handler.custom({
        entry: './myJsResolver.js', // relative or absolute path to 
        dataSource: a.ref('Post') // optional; can reference existing model or string data source; defaults to NONE_DS
      })
    ])
})

export const data = defineData({
  schema,
});
```

This definition returns the following IR from `data-schema` when calling `schema.transform()`:
```ts
{
  typeName: 'Mutation',
  fieldName: 'likePost',
  handlers: [
    {
      dataSource: 'PostTable',
      entry: { relativePath: './myJsResolver.js', importLine: `${stack_trace_import_line}` }
    }
  ]
}
```

The`convertJsResolverDefinition`function added in this PR consumes the IR and provisions an AppSync JS Pipeline Resolver and a list of functions.

## Checklist

- [x] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [x] If this PR requires a docs update, I have linked to that docs PR above.
- [x] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/backend-data/src/convert_js_resolvers.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+// import { JsResolver } from '@aws-amplify/data-schema-types';
+
+// Temporarily adding this type for draft PR. Will be imported from data-schema-types
+// once that package is updated with this new type
+export type JsResolver = {
+  typeName: 'Mutation' | 'Query' | 'Subscription';
+  fieldName: string;
+  handlers: {
+    dataSource: string;
+    entry: string;
```
Comment (edwardfoyle) [inline packages/backend-data/src/convert_js_resolvers.ts:15]: are we going to support any sort of build step for these files? or does it have to be raw js?

File: packages/backend-data/src/convert_js_resolvers.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  }[];
+};
+
+/**
+ * Default top-level pipeline resolver template
+ */
+const topLevelResolver = `
```
Comment (edwardfoyle) [inline packages/backend-data/src/convert_js_resolvers.ts:22]: maybe call this 'identityResolver' or 'passthroughResolver'?

Also, I think I would prefer putting this code in a separate file and reading it into a string here. That way we can at least have unit tests for these functions (even though they are trivial, code in template strings makes me nervous)

File: packages/backend-data/src/convert_js_resolvers.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        dataSourceName: handler.dataSource,
+        name: fnName,
+        // handler.entry is a resolved absolute path
+        code: readFileSync(handler.entry, 'utf-8'),
+        runtime: {
+          name: 'APPSYNC_JS',
+          runtimeVersion: '1.0.0',
+        },
+      });
+      fn.node.addDependency(amplifyApi);
```
Comment (edwardfoyle) [inline packages/backend-data/src/convert_js_resolvers.ts:58]: it should be okay to add these explicitly, but CFN should also be able to figure it out from the apiId reference in the resource config

**Review by edwardfoyle — COMMENTED** [review]
Overall I think this seems good. Could you put an example of what the DX in the backend definition looks like for thi?

File: packages/backend-data/src/convert_js_resolvers.ts (hunk 3 above)
Comment (iartemiev) [reply packages/backend-data/src/convert_js_resolvers.ts:58]: I was running into some sequencing issues without making the dep explicit, specifically, errors re: referencing nonexistent data sources. I think the apiId gets resolved before all the data sources are available.

File: packages/backend-data/src/convert_js_resolvers.ts (hunk 1 above)
Comment (iartemiev) [reply packages/backend-data/src/convert_js_resolvers.ts:15]: However, what should we keep in mind if we decide we want to add a build step in the future?

For example, we may want to ensure the `response` handler return type conforms to the return type configured in the schema definition. (We might look into this post-GA.)

File: packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/js_custom_fn.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable */
```
Comment (iartemiev) [inline packages/integration-tests/src/test-projects/data-storage-auth-with-triggers-ts/amplify/data/js_custom_fn.js:1]: Example JS resolver function used by e2e

**Comment by iartemiev** [comment]
Seeing
> Error: Could not determine how to assume credentials. Please check your inputs and try again.

After doing a bit of digging it looks like the PR has to be from a branch on upstream to run e2es. Re-opened here: https://github.com/aws-amplify/amplify-backend/pull/1089

**Pull request closed without merging**